new Pod Settings buttons to control and test beeping

itsmojo opened the pull request:
[Enable/Disable Confirmation Beeps] [Enable/Disable Optional Pod Alarms] [Check Beeps]
Optional cancelNone() which will verifies the nonce to replace getStatus(),
flash logs read support, beep on Pod startup completion to verify piezo

itsmojo committed:
new Pod Settings buttons & internal variables to control and test beeping:
[Enable/Disable Confirmation Beeps] [Enable/Disable Optional Pod Alarms] [Check Beeps]
Optional cancelNone() which will verifies the nonce to replace getStatus(),
flash logs read support, beep on Pod startup completion to verify piezo

itsmojo committed:
update some switch logic for new buttons

itsmojo committed:
Revert back to using hexWordLog for testing

ps2 commented on OmniKit/MessageTransport/MessageBlocks/BasalScheduleExtraCommand.swift:
```diff
@@ -62,7 +62,7 @@ public struct BasalScheduleExtraCommand : MessageBlock {
     }
     
-    public init(acknowledgementBeep: Bool, completionBeep: Bool, programReminderInterval: TimeInterval, currentEntryIndex: UInt8, remainingPulses: Double, delayUntilNextTenthOfPulse: TimeInterval, rateEntries: [RateEntry]) {
+    public init(acknowledgementBeep: Bool = false, completionBeep: Bool = false, programReminderInterval: TimeInterval = 0, currentEntryIndex: UInt8, remainingPulses: Double, delayUntilNextTenthOfPulse: TimeInterval, rateEntries: [RateEntry]) {
```
Parameters with defaults should go at the end of the function's parameter list.

itsmojo replied:
Fixed

ps2 commented on OmniKit/MessageTransport/MessageBlocks/PodInfoFlashLogRecent.swift:
```diff
@@ -8,24 +8,61 @@
 
 import Foundation
 
+let maxPumpEntriesReturned = 50
```
fileprivate

itsmojo replied:
Done

ps2 commented on OmniKit/Model/AlertSlot.swift:
```diff
@@ -99,13 +99,24 @@ public enum PodAlert: CustomStringConvertible, RawRepresentable, Equatable {
             return AlertConfiguration(alertType: .slot7, duration: .minutes(55), trigger: .timeUntilAlert(.minutes(5)), beepRepeat: .every5Minutes, beepType: .bipBeepBipBeepBipBeepBipBeep)
         case .expirationAlert(let alertTime):
-            return AlertConfiguration(alertType: .slot3, duration: 0, trigger: .timeUntilAlert(alertTime), beepRepeat: .every1MinuteFor3MinutesAndRepeatEvery15Minutes, beepType: .bipBeepBipBeepBipBeepBipBeep)
+            if alertTime == 0 {
```
```
let active = alertTime != 0 // disable if alertTime is 0
return AlertConfiguration(alertType: .slot3, active: active, duration: 0, trigger: .timeUntilAlert(alertTime), beepRepeat: .every1MinuteFor3MinutesAndRepeatEvery15Minutes, beepType: .bipBeepBipBeepBipBeepBipBeep)
```
This makes it clear how the check is being used and saves the reader looking through all the other arguments to see if there is anything else different.

itsmojo replied:
Done

ps2 commented on OmniKit/PumpManager/OmnipodPumpManager.swift:
```diff
@@ -31,6 +31,15 @@ public enum OmnipodPumpManagerError: Error {
     case notReadyForCannulaInsertion
 }
 
+public enum RunCommandSessionType {
```
I understand that you're trying to reduce the number of functions (and all the boiler plate that goes with them). But the savings comes at the expense of extensibility. If any of these commands later evolve to require parameters, or worse, want to return a different type, then it starts to get messy. I recommend keeping a function based api on the pump manager.

itsmojo replied:
Done, although I have some reservations about doing it this way.

ps2 commented on OmniKit/PumpManager/PodCommsSession.swift:
```diff
@@ -137,13 +137,21 @@ extension PodCommsError: LocalizedError {
 }
 
-
+public var confirmationBeeps: Bool = false // TODO determine placement, create methods to get/set, make persistent?
```
These belong on OmnipodPumpManagerState

itsmojo replied:
Not implemented due to lack of time

ps2 commented on OmniKitUI/ViewControllers/OmnipodSettingsViewController.swift:
```diff
@@ -282,6 +319,17 @@ class OmnipodSettingsViewController: RileyLinkSettingsViewController {
+                return cell
+            case .checkBeeps:
+                let cell = tableView.dequeueReusableCell(withIdentifier: SettingsTableViewCell.className, for: indexPath)
+                cell.textLabel?.text = LocalizedString("Check Beeps", comment: "The title of the command to run check beeps")
```
"Check Beeps" is somewhat ambiguous. My first assumption was that it reported the status of whether beeps were enabled. This should indicate it will play a sound on the pod.  Maybe "Emit Pod Beep", or "Play Test Beep"

itsmojo replied:
Check Beeps renamed to Play Test Beeps

ps2 reviewed (changes requested):
Thanks for submitting this early, Joe. Please see my comments on how we can get this merged. I'd be happy to take a look at the UI once the implementation detail comments have been addressed, if you wish.

itsmojo committed:
Various updates to address feedback received, "Check Beeps" is now "Play Test Beeps"

itsmojo committed:
Update for new BasalScheduleExtraCommand() parameter list

itsmojo committed:
Restructures to have comfirmationBeeps and optionalPodAlarms persistent
and included in OmnipodPumpManager state and not PodCommsSessions state

ps2 commented on OmniKit/PumpManager/OmnipodPumpManager.swift:
```diff
@@ -862,7 +887,121 @@ extension OmnipodPumpManager {
+    public func enableConfirmationBeeps(completion: @escaping (Error?) -> Void) {
+        self.confirmationBeeps = true
+        guard self.hasActivePod else {
+            completion(OmnipodPumpManagerError.noPodPaired)
```
Seems like this shouldn't be an error. It should be ok to enable beeps w/o a pod?

itsmojo replied:
{en,dis}ableConfirmationBeeps reworked to no longer report an error w/o a pod

ps2 commented on OmniKit/PumpManager/OmnipodPumpManager.swift:
```diff
@@ -862,7 +887,121 @@ extension OmnipodPumpManager {
+        }
+    }
+
+    public func disableConfirmationBeeps(completion: @escaping (Error?) -> Void) {
```
This function and the one above could be combined into a `setConfirmationBeeps(enabled: Bool)` function, since there are only two lines differing.

itsmojo replied:
Implemented in latest update

ps2 commented on OmniKit/PumpManager/OmnipodPumpManager.swift:
```diff
@@ -862,7 +887,121 @@ extension OmnipodPumpManager {
+        }
+    }
+
+    public func disableOptionalPodAlarms(completion: @escaping (Error?) -> Void) {
```
Maybe combine with above function?

itsmojo replied:
{enable,disable}OptionalPodAlarms are now combined into a single function with an argument

ps2 commented on OmniKit/PumpManager/PodCommsSession.swift:
```diff
@@ -481,6 +502,12 @@ public class PodCommsSession {
             podState.updateFromStatusResponse(status)
 
+            if deliveryType == .all && basalBeeps && confirmationBeeps {
+                // With a deliveryType of .all and a beepType other than .noBeep, the Omnipod will emit 3 beeps!
```
This function shouldn't create a new way of beeping, issuing another command that could fail. A successful suspend with a failed beep emit will mark this suspend as uncertain, which is not ok.
Considering this with the fact that this alternate way of issuing beeps is making the api confusing, I think this logic should be moved to the PumpManager layer, if it's kept at all. The primary use case this is serving is bolus confirmation. And I'd be happy to take a PR that just addressed that; would be a lot simpler, for sure!

itsmojo replied:
I think that we really should have a confirmation beep when suspending to match the one when resuming when Confirmation Beeps is enabled and basalBeeps is set. Unfortunately the Pod’s cancel delivery command can’t be used to do this as you’d get 3 sets of beeps in this case. I put this functionality into cancelDelivery() because it effectively implements what I think the Pod really should have done. For now decided to just remove this code and document as a TODO that suspending a Pod needs to have a single confirmation beep to match the one when resuming the Pod when Confirmation Beeps are enabled and basalBeeps have been configured

ps2 commented on OmniKit/PumpManager/OmnipodPumpManagerState.swift:
```diff
@@ -127,6 +133,18 @@ public struct OmnipodPumpManagerState: RawRepresentable, Equatable {
         }
+
+        if let rawConfirmationBeeps = rawValue["confirmationBeeps"] as? Bool {
+            self.confirmationBeeps = rawConfirmationBeeps
```
Maybe `self.confirmationBeeps = rawValue["confirmationBeeps"] as? Bool ?? false`?

itsmojo replied:
Done

ps2 commented on OmniKit/PumpManager/PodCommsSession.swift:
```diff
@@ -540,27 +567,90 @@ public class PodCommsSession {
+        var statusResponse: StatusResponse
+
+        if useCancelNoneForStatus {
+            statusResponse = try cancelNone() // handles update from status response
```
Any reason this path isn't calling `updateFromStatusResponse`?

itsmojo replied:
Yes and that is what the comment was explaining. I have since restructured this routine to make this even more obvious.

ps2 commented on OmniKit/PumpManager/PodCommsSession.swift:
```diff
@@ -326,8 +324,31 @@ public class PodCommsSession {
-    public func insertCannula() throws -> TimeInterval {
+
+    // emits the specified beep type and sets the completion beep flags based on the specified confirmationBeep value
+    public func emitBeep(beepType: BeepType, confirmationBeeps: Bool) throws {
```
The function's name should be changed to communicate that it is reconfiguring a setting on the pod.

itsmojo replied:
emitBeep() has been renamed to beepConfig()

ps2 commented on OmniKit/PumpManager/PodCommsSession.swift:
```diff
@@ -445,8 +466,8 @@ public class PodCommsSession {
     
-    public func cancelDelivery(deliveryType: CancelDeliveryCommand.DeliveryType, beepType: BeepType) -> CancelDeliveryResult {
-        
+    public func cancelDelivery(deliveryType: CancelDeliveryCommand.DeliveryType, beepType: BeepType, confirmationBeeps: Bool) -> CancelDeliveryResult {
```
This is a confusing api.  What does it mean to set beepType to `.none`, but confirmation beeps to `true`? Or with beepType set to some kind of beep, but confirmation beeps set to `false`?

itsmojo replied:
Agreed, but I blame a lot of this on the silly Pod cancelDelivery() semantics. beepType is the value to given to the Pod’s Cancel Delivery command and confirmationBeeps is needed both to indicate a confirmation beep is desired and to call beepConfig() to actually implement the confirmation beep for the deliveryType = .all case.

It the last this implementation (which has cancelDelivery providing a more logical cancel delivery for deliveryType: .all with confirmation beeps), when beepType is .none, confirmationBeeps is true, and deliveryType is .all, this would generate a single confirmation beep using the Pod beepConfig command. If beepType is some kind of beep, then you get whatever the Pod gives you and there will be no additional confirmation beep provided.

I have removed the confirmationBeeps argument to cancelDelivery() for now as I have decided to punt on optional resume delivery confirmation beeps for the time being to hopefully not delay things any further.

ps2 commented on OmniKit/PumpManager/PodCommsSession.swift:
```diff
@@ -606,6 +696,51 @@ public class PodCommsSession {
         transport.assertOnSessionQueue()
     }
+
+    public func enableConfirmationBeeps() throws {
```
Does these use cases warrant two new functions?

itsmojo replied:
{enable,disable}ConfirmationBeeps are now combined to setConfirmationBeeps() which takes an argument

ps2 reviewed (changes requested):
Thanks Joe. Definitely moving in the right direction. Main issue I see is the cancelDelivery adding another call that can fail and change dosing certainty in an invalid way.  Some other smaller changes for clarity.

itsmojo committed:
Various improvements and cleanup
+ better commenting and routine names for some odd Pod beep conditions
+ some code improvements for clarity and brevity
+ remove some now unneeded & unused routines
+ reworked so that the pump flash log can be read on a faulted Pod using Test Command if enabled
+ add new hasSetupCompletedPod and isSetupCompleted variables for faulted Pods
+ merge {enable,disable}ConfirmationBeeps() into new setConfirmationBeeps()
+ merge {enable,disable}OptionalPodAlarms() into new setOptionalPodAlarms()
+ use confirmation beeps for priming and cannula insertion if enabled
+ don't try to give confirmation beeps when suspending yet to avoid potential issues

itsmojo committed:
updated for BolusExtraCommand() argment order

itsmojo committed:
Merge branch 'dev' of https://github.com/ps2/rileylink_ios into more-pod-settings-options

Merge lastest dev code to hopefully pickup up testing builds fixes

itsmojo committed:
Fix totally broken test for PodInfoFlashLogRecent
which included a duplicate packet and was missing items in another list
Add new test for PodInfoFlashLogPrevious from the same Pod run

itsmojo committed:
acknowledgementBeep and confirmationBeep don't work for priming and insert cannula
updated comments and some minor code improvements for beep handling and faults

itsmojo committed:
Use new BeepConfigType for $1E Beep Config and BeepType for $19 Configure Alerts and $1F Cancel
Have Disable Confirmation Beeps use BeepConfigType .beepConfig_NoBeep to silently disable completion beeps

ps2 commented on OmniKit/PumpManager/OmnipodPumpManager.swift:
```diff
@@ -862,7 +893,91 @@ extension OmnipodPumpManager {
+    }
+
+    public func setConfirmationBeeps(enabled: Bool, completion: @escaping (Error?) -> Void) {
+        self.confirmationBeeps = enabled
```
If this command fails, it will report an error to the user, but will still have confirmation beeps enabled.

ps2 commented on OmniKit/PumpManager/OmnipodPumpManager.swift:
```diff
@@ -862,7 +893,91 @@ extension OmnipodPumpManager {
+    }
+
+    public func setOptionalPodAlarms(enabled: Bool, completion: @escaping (Error?) -> Void) {
+        self.optionalPodAlarms = enabled
```
Similarly to setConfirmationBeeps, we should probably only set this if we return a success condition.

ps2 commented on OmniKit/PumpManager/PodCommsSession.swift:
```diff
@@ -326,8 +324,34 @@ public class PodCommsSession {
-    public func insertCannula() throws -> TimeInterval {
+
+    // emits the specified beep type and sets the completion beep flags based on the specified confirmationBeep value
+    public func beepConfig(beepConfigType: BeepConfigType, confirmationBeeps: Bool) throws {
```
This function should take `basalBeeps`, `tempBasalBeeps`, and `bolusBeeps` as arguments, instead of referring to global variables.

ps2 commented on OmniKit/PumpManager/PodCommsSession.swift:
```diff
@@ -326,8 +324,34 @@ public class PodCommsSession {
+
+        if optionalPodAlarms {
+            // program optional Pod alarms, will beep upon completion if supplementaryBeeps && confirmationBeeps
+            try enableOptionalPodAlarms(confirmationBeeps: confirmationBeeps)
```
If these fail, the exceptions will be thrown back to insertCannula, and reported as an error to the user, even though the setupProgress is .completed.  The UI will offer a retry to the user, which will then fail, as the pod is not in the right state.  If you want to consider this successful, I'd recommend swallowing the exceptions for the beeps here.

ps2 commented on OmniKit/Model/BeepType.swift:
```diff
@@ -19,10 +19,24 @@ public enum BeepType: UInt8 {
-    case fiveSecondBeep = 0xE    // 5 second alarm beep *if* Pod is currently suspended
-    case beepConfig_NoBeep = 0xF // For Beep Config Command no beep, else fatal Pod fault $37
+    case fiveSecondBeep = 0xE    // can only be used if Pod is currently suspended
+    case beepConfig_NoBeep = 0xF
```
`beepConfig_NoBeep` -> `noBeep`

ps2 reviewed (commented):
The global variables `basalBeeps`, `tempBasalBeeps` and `bolusBeeps` are being used a lot of different code paths, and conditionals. I'm not comfortable having such a wide reaching set of effects added to the code using global variables.  If these are features we want to support, then they need to be added in a maintainable manner.

itsmojo committed:
Proposed version with all of the let constants to customze the ConfirmationBeep behavior removed
Have noBeep appear on both BeepType and BeepConfigType enums but with a different value
Move playTestBeeps out of PodCommSession (which makes it unavailable for testingCommands() :-()
Rework beepConfig() to catch any exceptions
Don't set optionalPodAlarms variable until after the successful configuration of the Pod alerts
Implemented reasonable confirmation beep behavior for Change Time Zone and Suspend Delivery
Mark setupProgress as complete after any optional Pod alarms successfully set up
Buttons reworked to use isLoading while in progress for a spinner and then reset state when done

ps2 commented:
Work ongoing in separate PR
